R/cube-result.R

```diff
@@ -434,3 +402,147 @@ setMethod("bases", "CrunchCube", function (x, margin=NULL) {
         return(cubeMarginTable(x, margin, measure=".unweighted_counts"))
     }
 })
+
+
+
+#' Work with CrunchCubes, MultitableResults, and TabBookResults
+#'
+#' These functions provide an interface like [base::margin.table()] and
+#' [base::prop.table()] for the CrunchCube object. CrunchCubes contain richer
+#' metadata than standard R `array` objects, and they also conceal certain
+#' complexity in the data structures from the user. In particular,
+#' multiple-response variables are generally represented as single dimensions in
+#' result tables, but in the actual data, they may comprise two dimensions.
+#' These methods understand the subtleties in the Crunch data types and
+#' correctly compute margins and percentages off of them.
+#'
+#' These functions also generalize to MultitableResults and TabBookResults,
+#' which are returned from a [tabBook()] request. When called on one of those
+#' objects, they effectively apply over each CrunchCube contained in them.
+#'
+#' `bases` is an additional method for CrunchCubes. When making weighted
+#' requests, `bases` allows you to access the unweighted counts for every cell
+#' in the resulting table (array). The `bases` function takes a "margin"
+#' argument to work like `margin.table`, or with `margin=0` gives all cell
+#' counts.
+#'
+#' `collapse.dimensions` returns a cube that collapses the dimensions given in
+#' `margin`. This is useful if you want to get counts that are equivalent to a
+#' univariate cube from a multivariate cube. For example
+#' `collapse.dimensions(crtabs(~ fruit + pets, ds), 2)` will be equal to
+#' `crtabs(~ fruit, ds)` and `collapse.dimensions(crtabs(~ fruit + pets, ds),
+#' 1)` will be equal to `crtabs(~ pets, ds)`.
```

Comment:
Perhaps note that for the simple (non-MR) case, this is equivalent to `apply(x, margin, sum)`. Maybe that also helps with the function name, which I don't find particularly obvious. It's a bit cheeky, but the generalization of `rowSums()` and `colSums()` might be `dimSums()` ;)

Also, in what way are they "equal"? Are they identical cube objects? Have equivalent counts/margins/percentages/etc. based on the supported methods?

Reply:
Hah, I like `dimSums()`, I'll `s/collapse.dimensions/dimSums/g` and coalesce the arguments if i can.

The cube objects are identical, though like you point out below, that's only really true for count measures.